File: content/problems/0934-shortest-bridge/solutions/island-marking-multisource-bfs/Solution.java
class Solution {
    public int shortestBridge(int[][] grid){
        int n=grid.length;
        Deque<int[]> flood=new ArrayDeque<>(),frontier=new ArrayDeque<>();
        outer:for(int r=0;r<n;r++)for(int c=0;c<n;c++)if(grid[r][c]==1){
            flood.offer(new int[]{
                r,c
            });
            grid[r][c]=2;
            break outer;
        }
        int[][] directions={
            {
                1,0
            },{
                -1,0
            },{
                0,1
            },{
                0,-1
            }
        };
        while(!flood.isEmpty()){
            int[] cell=flood.poll();
            frontier.offer(cell);
            for(int[] d:directions){
                int r=cell[0]+d[0],c=cell[1]+d[1];
                if(r>=0&&r<n&&c>=0&&c<n&&grid[r][c]==1){
                    grid[r][c]=2;
                    flood.offer(new int[]{
                        r,c
                    });
                }
            }
        }
        int distance=0;
        while(!frontier.isEmpty()){
            for(int size=frontier.size();size>0;size--){
                int[] cell=frontier.poll();
                for(int[] d:directions){
                    int r=cell[0]+d[0],c=cell[1]+d[1];
                    if(r<0||r>=n||c<0||c>=n||grid[r][c]==2)continue;
                    if(grid[r][c]==1)return distance;
                    grid[r][c]=2;
                    frontier.offer(new int[]{
                        r,c
                    });
                }
            }
            distance++;
        }
        return -1;
    }
}

File: content/problems/0934-shortest-bridge/solutions/island-marking-multisource-bfs/Solution.kt
class Solution {
    fun shortestBridge(grid:Array<IntArray>):Int{
        val n=grid.size;
        val flood=java.util.ArrayDeque<IntArray>();
        val frontier=java.util.ArrayDeque<IntArray>();
        loop@for(r in 0 until n)for(c in 0 until n)if(grid[r][c]==1){
            flood.addLast(intArrayOf(r,c));
            grid[r][c]=2;
            break@loop
        };
        val directions=arrayOf(intArrayOf(1,0),intArrayOf(-1,0),intArrayOf(0,1),intArrayOf(0,-1));
        while(flood.isNotEmpty()){
            val cell=flood.removeFirst();
            frontier.addLast(cell);
            for(d in directions){
                val r=cell[0]+d[0];
                val c=cell[1]+d[1];
                if(r in 0 until n&&c in 0 until n&&grid[r][c]==1){
                    grid[r][c]=2;
                    flood.addLast(intArrayOf(r,c))
                }
            }
        };
        var distance=0;
        while(frontier.isNotEmpty()){
            repeat(frontier.size){
                val cell=frontier.removeFirst();
                for(d in directions){
                    val r=cell[0]+d[0];
                    val c=cell[1]+d[1];
                    if(r !in 0 until n||c !in 0 until n||grid[r][c]==2)continue;
                    if(grid[r][c]==1)return distance;
                    grid[r][c]=2;
                    frontier.addLast(intArrayOf(r,c))
                }
            };
            distance++
        };
        return -1
    }
}
